bin/annotate_test_failures

```diff
@@ -17,28 +17,46 @@ end
 file = File.open(junit_path)
 xmldoc = REXML::Document.new(file)
 
+# Here's how report.junit XML files look like for failure vs success:
+#
+# <testcase classname='WordPressTest.MediaServiceTests' name='testDeletingLocalMediaThatDoesntExistInCoreData'>
+#   <failure message='Asynchronous wait failed: Exceeded timeout of 0.1 seconds, with unfulfilled expectations: &quot;The delete call succeeds even if the media object isn&apos;t saved.&quot;.'>&lt;unknown&gt;:0</failure>
+# </testcase>
+# <testcase classname='WordPressTest.MediaServiceTests' name='testDeletingLocalMediaThatDoesntExistInCoreData' time='0.145'/>
+#
 failures = []
 REXML::XPath.each(xmldoc, '//testcase[failure]') do |testcase|
-  failure = testcase.elements['failure']
-  failures.append <<~ENTRY
-  <details><summary><tt>#{testcase['name']}</tt> in <tt>#{testcase['classname']}</tt></summary>
-  
-  #{failure['message']}
+  test_class = testcase['classname']
+  test_name = testcase['name']
+  failure_node = testcase.elements['failure']
+  failure_message = failure_node['message']
 
-  ```
-  #{failure.text}
-  ```
-  </details>
-  ENTRY
+  # Find all nodes that are about the same test, to detect false positives
+  # (in case there were multiple retries of the same flaky test with the last one ultimately passing)
+  retries = testcase.parent.get_elements("testcase[@classname='#{testcase['classname']}' and @name='#{testcase['name']}']")
+  if retries.last.elements['failure']
+    failures.append <<~ENTRY
+      <details><summary><tt>#{test_name}</tt> in <tt>#{test_class}</tt></summary>
+      #{failure_message}
+  
+      ```
+      #{failure_node.text}
+      ```
+      </details>
+    ENTRY
+  else
```

Comment:
Side note that if a test case was retried, say, 3 times, but failed in all 3 cases (so, _not_ ultimately succeeding), then all 3 failures for that same test will be reported (as opposed to, say, the last failure only)

This is intentional, because it's not so uncommon that flaky tests fail at different points in the test and/or with different error messages during their various retries.

<details><summary>That's actually exactly what happened for <code>testDeleteFollowerFailure</code> in <a href="https://buildkite.com/automattic/wordpress-ios/builds/12940#0186c2cf-d9f3-440b-b8db-67ca28a271a9">this CI build</a></summary>
<img width="705" alt="image" src="https://user-images.githubusercontent.com/216089/224164131-e8c1b244-2cc1-461d-9f0e-d561186dc133.png">

Even if, in this specific case, it ultimately passed on 4th retry (which is why the whole CI step ended up green, and you can also see that by looking in the `report.junit`), you can imagine a case where it wouldn't have ultimately succeeded, yet it'd still be interesting to see the 3 different error messages for each failure instead of only the last one. Especially to avoid accidentally misleading the person who will try to debug that test failure into thinking that the test always fails at the same point in code and with the same error, when the reality is that it's happening more randomly.
</details>

Reply:
> Side note that if a test case was retried, say, 3 times, but failed in all 3 attempts (so, not ultimately succeeding), then all 3 failures for that same test will be reported (as opposed to, say, the last failure only)
>
> This is intentional, because it's not so uncommon that flaky tests fail at different points in the test and/or with different error messages during their various retries.

Are we sure that the number of `testcase.failure` nodes is the number of retries?

WordPress iOS is configured to retry on failure up to 3 times, but `testDeleteFollowerFailure` is not marked as a consistent failure despite having 3 failures

<img width="534" alt="image" src="https://user-images.githubusercontent.com/1218433/224225587-b3d736ba-2477-4430-a78d-a73026650959.png">

_retry on failure, maximum three times_

<img width="1171" alt="image" src="https://user-images.githubusercontent.com/1218433/224225782-b2e025f7-b4b7-48b3-8cc3-4e7414d5f612.png">

_3 failure, 1 success... 4 retries?_ 🤔 

This _could_ be a bug in Xcode, but [the way `testDeleteFollowerFailure` is written](https://github.com/wordpress-mobile/WordPress-iOS/blob/e59cda935c7e02b68c8c03afcbcf2cb9b58e6728/WordPress/WordPressTest/PeopleServiceTests.swift#L154-L203) also leaves the door open for the test reporting multiple failures in a single run. After all, it has multiple assertions.

In particular, notice how a failure in the Nimble async code doesn't prevent the test from finishing running and failing again later on (the async code is just how the test is shaped, Nimble `waitUntil` ... `done()` makes the execution "linear"):

<img width="1719" alt="image" src="https://user-images.githubusercontent.com/1218433/224226559-e5a78504-61dc-4e17-b797-29f685ca063d.png">

This is, as far as I know, the expected behavior. Here's a dumb example:

<img width="1346" alt="image" src="https://user-images.githubusercontent.com/1218433/224227005-75b2bf0a-a3ac-4225-aa52-c5b37e7ecfef.png">

Reply:
I run those tests locally, this was the JUnit output:

```
    <testcase classname='WordPressTest.PeopleServiceTests' name='testMultipleFailuresInOneTest'>
      <failure message='XCTAssertTrue failed'>WordPress/WordPressTest/PeopleServiceTests.swift:226</failure>
    </testcase>
    <testcase classname='WordPressTest.PeopleServiceTests' name='testMultipleFailuresInOneTest'>
      <failure message='XCTAssertFalse failed'>WordPress/WordPressTest/PeopleServiceTests.swift:227</failure>
    </testcase>
    <testcase classname='WordPressTest.PeopleServiceTests' name='testMultipleFailuresInOneTest'>
      <failure message='XCTAssertTrue failed'>WordPress/WordPressTest/PeopleServiceTests.swift:226</failure>
    </testcase>
    <testcase classname='WordPressTest.PeopleServiceTests' name='testMultipleFailuresInOneTest'>
      <failure message='XCTAssertFalse failed'>WordPress/WordPressTest/PeopleServiceTests.swift:227</failure>
    </testcase>
    <testcase classname='WordPressTest.PeopleServiceTests' name='testMultipleFailuresInOneTest'>
      <failure message='XCTAssertTrue failed'>WordPress/WordPressTest/PeopleServiceTests.swift:226</failure>
    </testcase>
    <testcase classname='WordPressTest.PeopleServiceTests' name='testMultipleFailuresInOneTest'>
      <failure message='XCTAssertFalse failed'>WordPress/WordPressTest/PeopleServiceTests.swift:227</failure>
    </testcase>
```

That refutes the `testcase.failure` count equals retries hypothesis 😞

Reply:
Running that test report through the changes in this PR (as of https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/46/commits/0727e65e25415f5f1f13bd0255edb6deceb0f3c7) results in the following output

---

Ultimately found 6 real test failure(s). Reporting them as a `Tests` Buildkite annotation.
#### Tests: 6 failure(s)

<details><summary><tt>testMultipleFailuresInOneTest</tt> in <tt>WordPressTest.PeopleServiceTests</tt></summary>
XCTAssertTrue failed

```
WordPress/WordPressTest/PeopleServiceTests.swift:225
```
</details>

<details><summary><tt>testMultipleFailuresInOneTest</tt> in <tt>WordPressTest.PeopleServiceTests</tt></summary>
XCTAssertFalse failed

```
WordPress/WordPressTest/PeopleServiceTests.swift:226
```
</details>

<details><summary><tt>testMultipleFailuresInOneTest</tt> in <tt>WordPressTest.PeopleServiceTests</tt></summary>
XCTAssertTrue failed

```
WordPress/WordPressTest/PeopleServiceTests.swift:225
```
</details>

<details><summary><tt>testMultipleFailuresInOneTest</tt> in <tt>WordPressTest.PeopleServiceTests</tt></summary>
XCTAssertFalse failed

```
WordPress/WordPressTest/PeopleServiceTests.swift:226
```
</details>

<details><summary><tt>testMultipleFailuresInOneTest</tt> in <tt>WordPressTest.PeopleServiceTests</tt></summary>
XCTAssertTrue failed

```
WordPress/WordPressTest/PeopleServiceTests.swift:225
```
</details>

<details><summary><tt>testMultipleFailuresInOneTest</tt> in <tt>WordPressTest.PeopleServiceTests</tt></summary>
XCTAssertFalse failed

```
WordPress/WordPressTest/PeopleServiceTests.swift:226
```
</details>

Reply:
I'm not sure where to go from here... I guess we could change the way tests are grouped for the purpose of reporting?

Xcode knows how to differentiate between multiple assertion failures in the same `func test...` and a whole `func test...` failing:

![image](https://user-images.githubusercontent.com/1218433/224229599-10bbf6fc-df71-43dc-b0a1-90ff71b77581.png)

That makes me think the `xcresult` output would contain that information, too, and that it is the translation to JUnit XML format that looses it. But, I haven't had a chance to verify it.

Reply:
Ahhh damn, you make a good point in the fact that a given test case can have multiple assertion failures surfaced in a single execution… didn't think of that.

That kinda throws a wrench in the whole logic of my implementation… or does it? Because even if a single run can have multiple failures, would it then end up having a non-failing XML node at the end if the test didn't pass?

In other words, yes, if a test case reports 3 failure nodes in the junit report, maybe indeed it doesn't mean 3 retries (each failing once), but instead 3 failures reported in the first time running the test. But even then, I wouldn't expect a `<testcase>` node _without_ a `<failure>` child node… showing up at the end of the list, unless if the test ultimately succeeded. Because while a `XCTAssert` failing would call `XCTestCase.recordFailure(…)` and ultimately report a failure entry in the `xcresult` and `.junit` reports… a non-failing `XCTAssert` assertion does _not_ report a success entry (there's no underlying `XCTestCase.recordSuccess(…)`) on its own.

So I think that, while there are multiple possible scenarios that would lead to having multiple failure nodes (either multiple failures in a single run, or failures happening during multiple retries, or a mix of both), the only way to get a _success_ node (aka a `testcase` node without a child `failure` node) for a test in the `junit` report is if the whole test ultimately passed?

Reply:
Said otherwise, even if I agree that the "`testcase.failure` count equals retries" hypothesis is probably wrong after all, the "test case ends up with a success means Xcode ultimately considered it green and marked as passed" hypothesis still holds, making the logic still work (even if we should maybe rename the `retries` variable in the implementation to something like `entries_for_same_test` or similar to avoid misleading the reader)?

---

My new theory is thus that, given the following:

```swift
func testFlaky() {
  XCTAssertTrue(Bool.random(), message: "1")
  XCTAssertTrue(Bool.random(), message: "2")
}
```
And Xcode configured for 3 retries:
 - first try failing in both assertions, second try failing only on `2`, but third try passing, will lead to nodes representing {`failure:1`, `failure:2`,`failure:2`,`success`}
 - first and second try failing on `2`, third and last try failing on `1`, and the test never succeeding despite retries… will lead to {`failure:2`,`failure:2`,`failure:1`}, with no success node.

Still making the "if set ends with success node, then whole test was ultimately marked green, and if not then it was red, whether that was due to multiple retries or multiple assertions in single try or mix of both" logic work…?

Reply:
> "if set ends with success node, then whole test was ultimately marked green, and if not then it was red, whether that was due to multiple retries or multiple assertions in single try or mix of both" logic work…?

Yes, I think that rule would be correct under the assumption that the test suite either has test repetition set as _None_ or _Retry on Failure_.

It might break if we used _Until Failure_ or _Up Until Maximum Repetition_ depending on what one considers a success, give the set might end with a failure in those cases even though previous repetitions were successful. 🤔 Personally, I think both options are useful only for stress testing and given we are not doing any of that, I don't think we should worry about it too much for the moment.